dashboard/config/custom_session_store/instrumented_cookie_store.rb

```diff
@@ -0,0 +1,45 @@
+require 'action_dispatch/middleware/session/cookie_store'
+require 'objspace'
+
+module ActionDispatch
+  module Session
+    # Simple wrapper around the standard Rails decentralized cookie-based
+    # session store that we currently use in production. The primary goal is to
+    # collect some data about current usage patterns, to inform our decisions
+    # about a possible switch to a centralized session store.
+    #
+    # TODO infra: remove this helper after we've gathered enough data.
+    class InstrumentedCookieStore < CookieStore
+      METRICS_NAMESPACE = 'code-dot-org/ActionDispatch'.freeze
+      METRICS_LOG_WINDOW = 1.minute
+
+      # @see https://github.com/rails/rails/blob/ac87f58207cff18880593263be9d83456aa3a2ef/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb#L104-L107
+      private def write_session(req, session_id, session_data, options)
+        log_session_info(session_id, session_data)
+        super(req, session_id, session_data, options)
+      end
+
+      # Custom helper method to log some data about active sessions.
+      # Specifically, this helper will record the IDs and sizes of all sessions
+      # seen during a configurable window of time, and at the end of that
+      # window will log to CloudWatch some data summarizing all sessions seen
+      # during the window.
+      private def log_session_info(session_id, session_data)
+        @session_info ||= {}
+        @session_info_last_logged_at ||= Time.now
+
+        session_size = ObjectSpace.memsize_of(session_data)
+        @session_info[session_id.private_id] = [session_size, @session_info.fetch(session_id.private_id, 0)].max
+
+        if @session_info_last_logged_at < Time.now - METRICS_LOG_WINDOW
+          num_sessions = @session_info.size
+          Cdo::Metrics.put(METRICS_NAMESPACE, 'num_sessions_seen', num_sessions, {Environment: CDO.rack_env})
```

Comment:
Something I considered is actually adding the configured window length as a dimension, so if at some point in the future we want to change the frequency of logging, we won't be confused by the dramatic change in the number of sessions seen during that window.

```suggestion
          Cdo::Metrics.put(METRICS_NAMESPACE, 'num_sessions_seen', num_sessions, {Environment: CDO.rack_env, MetricsLogWindow: METRICS_LOG_WINDOW})
```

I ultimately didn't bother because it felt a bit unnecessary; let me know if you think it could be valuable!

Reply:
I'm betting on us getting enough info from one sampling, don't overbuilds it says I.